### PR TITLE
Follower cursor implementation

### PR DESCRIPTION
**Pull request by merlimat**
This is stacked on top of #54 

The `FollowerCursor` is created by the leader controller for each follower and it handles pushing the updates and receiving the acks back, passing them to the `AckQuorumTracker`.

### CODE REVIEW

File: server/wal/wal_tidwall.go [hunk 1]
```diff
@@ -293,7 +293,8 @@ func (r *forwardReader) HasNext() bool {
 	if r.closed {
 		return false
 	}
-	return r.nextOffset <= r.maxIdInclusive.Offset && EntryIdFromProto(r.maxIdInclusive) != EntryId{}
+
+	return r.nextOffset <= r.wal.lastEntryId.Offset && EntryIdFromProto(r.maxIdInclusive) != EntryId{}
```
Comment (andrasbeni) [inline server/wal/wal_tidwall.go:297]: Please update the documentation comment of `HasNext()` accordingly.

File: server/follower_cursor_test.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	"testing"
+	"time"
+)
+
+func TestFollowerCursor(t *testing.T) {
+	var epoch uint64 = 1
+	var shard uint32 = 2
+
+	stream := newMockClientAddEntriesStream()
+	ackTracker := NewQuorumAckTracker(3, wal.EntryId{Epoch: 0, Offset: 0})
```
Comment (andrasbeni) [inline server/follower_cursor_test.go:16]:
```suggestion
	ackTracker := NewQuorumAckTracker(3, wal.NonExistentEntryId)
```

Comment (andrasbeni) [reply server/follower_cursor_test.go:16]: Maybe we should rename it to `wal.LogStartEntryId`?

File: server/wal/wal_tidwall.go (hunk 1 above)
Comment (merlimat) [reply server/wal/wal_tidwall.go:297]: The problem is that it's inconsistent with `ReadNext()`: 
 * ReadNext max offset is updated by reading the new value from the channel
 * HasNext is not updated

Comment (merlimat) [reply server/wal/wal_tidwall.go:297]: I don't think we need to keep track of the readers and have this channel mechanism to notify them. Each reader can just read the variable of what is the latest entry committed in the wal.

File: server/follower_cursor_test.go (hunk 2 above)
Comment (merlimat) [reply server/follower_cursor_test.go:16]: I plan to use `int64` for offset anyway and use `-1` for all these scenarios

File: server/wal/wal_tidwall.go (hunk 1 above)
Comment (merlimat) [reply server/wal/wal_tidwall.go:297]: (updated the doc for now)

File: server/follower_cursor_test.go (hunk 2 above)
Comment (andrasbeni) [reply server/follower_cursor_test.go:16]: That makes sense.

File: server/wal/wal_tidwall.go (hunk 1 above)
Comment (andrasbeni) [reply server/wal/wal_tidwall.go:297]: Yes, I believe this series of pull requests will make a lot of what is there now redundant.

**Pull request merged**